Alexander-M-rss/nodejs-aws-shop-be
# Task 10 - Backend For Frontend

1. Task: https://github.com/rolling-scopes-school/aws/blob/main/aws-developer/10_backend_for_frontend/task.md
2. Deploy:
> - Product Service API URL -  https://7uxztr6cn6.execute-api.eu-west-1.amazonaws.com/products
> - Cart Service API URL - http://alexander-m-rss-cart-api-prod.eu-west-1.elasticbeanstalk.com/api/profile/cart
https proxied link - https://11077afyee.execute-api.eu-west-1.amazonaws.com/prod/api/profile/cart
> - BFF Service API URL - http://alexander-m-rss-bff-api-prod.eu-west-1.elasticbeanstalk.com
https proxied link - https://prm9wvc3z6.execute-api.eu-west-1.amazonaws.com/prod
> - FE application - https://d2tbewbqvvmzls.cloudfront.net/ - works via https proxied BFF link
http deploy - http://rss-my-shop-http-deploy.s3-website-eu-west-1.amazonaws.com/ - uses direct BFF link
_To avoid mixed-content blocking with direct BFF link http deploy from s3-website is provided_
> - example of how to call `createProduct` lambda with all needed information: URL, payload, headers, etc.
>> - product-service
>>```bash
>> curl --location 'https://7uxztr6cn6.execute-api.eu-west-1.amazonaws.com/products' \
>> --header 'Content-Type: application/json' \
>> --data '{
>>   "title": "Product title",
>>   "description": "Product description",
>>   "price": 7,
>>   "count": 10,
>>   "image_url": "https://images.unsplash.com/photo-1689442529342-3f1c7513d9ae"
>> }'
>>```
>> _`image_url` is optional_
>> - bff-service
>>```bash
>> curl --location 'http://alexander-m-rss-bff-api-prod.eu-west-1.elasticbeanstalk.com/products' \
>> --header 'Content-Type: application/json' \
>> --data '{
>>   "title": "Product title",
>>   "description": "Product description",
>>   "price": 7,
>>   "count": 10,
>>   "image_url": "https://images.unsplash.com/photo-1689442529342-3f1c7513d9ae"
>> }'
>>```
>> _`image_url` is optional_
> - example how to call Product Service and Cart Service via BFF Service URL
>> - Product service - http://alexander-m-rss-bff-api-prod.eu-west-1.elasticbeanstalk.com/products
proxied - https://prm9wvc3z6.execute-api.eu-west-1.amazonaws.com/prod/products
>> - Cart service - http://alexander-m-rss-bff-api-prod.eu-west-1.elasticbeanstalk.com/cart
proxied - https://prm9wvc3z6.execute-api.eu-west-1.amazonaws.com/prod/cart
>> 
>> For Cart Service use valid Basic authorization token any username and password in Authorization header.
>> You can use this [**postman collection**](https://api.postman.com/collections/19569509-7b617e67-1151-4fee-9354-7f325f70c31b?access_key=PMAT-01H63R5G8M1F1Y6SAG6GXTC3ZE) **after adding your token**

3. Deadline: 25.07.2023
4.  FE PR https://github.com/Alexander-M-rss/nodejs-aws-shop-react/pull/8
5. Score: **100/100**
> * [X] `+80` Main task
>> * [X]  - A working and correct application should be in the `bff-service` folder. Reviewer can start this application locally with any valid configuration in the `.env` file and this application should works as described in the Task 10.1
>> * [X]- The BFF Service should be deployed with Elastic Beanstalk. The BFF Service call should be redirected to the appropriate service : Product Service or Cart Service. The response from the BFF Service should be the same as if Product Service or Cart Service services were called directly.
>> ## Application Functionality (**MUST HAVE**)
>>> * [x] 1. Products representation on Home page should be based on Product Service API.
>>> * [x] 2. Products are coming from Product DB.
>>> * [x] 3. Product images are not randomly generated on client side. Product image, same as another product model information should be stored on BE side in Product DB.
>>> * [x] 4. Products might be created through CSV product file import from client side. **New products may not be displayed for 2 minutes due to caching requests to the `getProductsList` lambda function. It depends on the time of the last request to  to the `getProductsList` lambda function.**
>>>> Please use following format for csv file:
>>>> ![csv](https://github.com/Alexander-M-rss/nodejs-aws-shop-be/assets/96181572/7a17cd5e-1339-49b6-8112-8ed6d6c0a916)
>>>>
>>>>![csv-txt](https://github.com/Alexander-M-rss/nodejs-aws-shop-be/assets/96181572/78ac7731-fd66-4c2f-bdaf-29d344f756e3)
>>>> **Attention!  _image_url_ value is optional. But if you specify _image_url_ value please use valid URL. If _image_url_ value is not valid URL, the product will not be created.**
>>> * [x] 5. Cart might be created with appropriate product set.
>>> * [x] 6. Auth logic should be in place
>>>> - **For `import-service` use token as desribed  [Task 7](https://github.com/rolling-scopes-school/aws/blob/main/aws-developer/07_authorization/task.md)**
>>>> **Note: ``Hyphens`` are __not allowed__ in environment variable names. So please use `underscores` instead of `hyphens` or skip them in the login. For example, if you see `My-Login` then use `my_login` or `mylogin`. (Two users are created. ) The login is case insensitive like the GitHub login. On client startup an empty token variable is created if it's not in localStorage for your convenience, you will only have to edit its value. If you have trouble generating a valid token, please contact me on discord or in the school app.**
>>>> - **For `cart-service` use  valid Basic authorization token any username and password in Authorization header. _Please don't use token from `import-service` to avoid multiple clients accessing the shopping cart at the same time!_**
>>>>
>>>>**On FE client startup an empty token variable is created if it's not in localStorage for your convenience, you will only have to edit its value. If you have trouble generating a valid token, please contact me on discord or in the school app.**
> * [x] `+20` Additional task
>> * [x] `+20` - Add a cache at the BFF Service level for a request to the `getProductsList` lambda function of the Product Service. The cache should expire in 2 minutes.  
>>   How to test:
>>   - Get products list
>>   - Create new product
>>   - Get products list - result shouldn’t have new product
>>   - Wait more than 2 minutes
>>   - Get products list - result should have new product
